File: backend/app/schemas/policy_search.py
from typing import Any, Literal
from pydantic import BaseModel, Field, ConfigDict
from app.schemas.policy import (
    ApplicationStatus,
    PolicyCategory,
    PolicyRead,
    PolicySort,
)

SearchDimension = Literal["keyword", "region", "age", "category", "status"]


class PolicySearchQueryParams(BaseModel):
    q: str = Field(
        default="",
        description=(
            "자연어 검색어 (권장 최대 200자). q가 비어 있으면 region, age, "
            "category, status, keyword 중 하나 이상의 명시 조건이 필요함"
        ),
        max_length=200,
    )
    keyword: str | None = Field(
        default=None,
        description="명시적 키워드 필터 (q의 키워드 조건 override, 권장 최대 100자)",
        max_length=100,
    )
    region: str | None = Field(
        default=None,
        description="명시적 지역 alias/name 문자열 (q의 지역 조건 override, 권장 최대 100자)",
        max_length=100,
    )
    age: int | None = Field(
        default=None,
        ge=0,
        le=150,
        description="명시적 만 연령 (q의 연령 조건 override)",
    )
    category: PolicyCategory | None = Field(
        default=None,
        description="명시적 정책 카테고리 (q의 카테고리 override)",
    )
    status: ApplicationStatus | None = Field(
        default=None,
        description="신청 상태 필터 (open, scheduled, closed)",
    )
    include_partial: bool = Field(
        default=True,
        description="partial 데이터 포함 여부 (검색 API 기본값: true)",
    )
    page: int = Field(default=1, ge=1, description="페이지 번호")
    limit: int = Field(default=20, ge=1, le=100, description="페이지당 결과 수")
    sort: PolicySort = Field(
        default="default",
        description=(
            "정렬: default(관련도), title_asc, title_desc, deadline_asc, "
            "deadline_desc, collected_desc, collected_asc"
        ),
    )


class PolicySearchPreferences(BaseModel):
    region: str | None = Field(default=None, max_length=100)
    age: int | None = Field(default=None, ge=0, le=150)
    categories: list[PolicyCategory] = Field(
        default_factory=list,
        max_length=7,
    )


class PolicySearchPostRequest(PolicySearchQueryParams):
    preferences: PolicySearchPreferences | None = None


class ConditionItem(BaseModel):
    dimension: SearchDimension = Field(..., description="해석된 차원 종류")
    value: Any = Field(..., description="추출 또는 명시 지정된 차원 값")
    source: Literal["q", "explicit"] = Field(
        ..., description="조건 출처 (q: 자연어 파싱, explicit: 명시적 쿼리 파라미터)"
    )
    resolution: Literal["resolved", "unmapped", "ambiguous"] = Field(
        ..., description="해석 결과 상태"
    )
    candidates: list[str] = Field(
        default_factory=list, description="매핑 후보 문자열 리스트 (ambiguous/unmapped 시 대안 제공)"
    )


class InterpretedConditions(BaseModel):
    q_raw: str = Field(..., description="사용자가 입력한 원본 자연어 검색어")
    q_clean: str = Field(..., description="전처리 및 공백 정리된 검색어")
    conditions: list[ConditionItem] = Field(
        default_factory=list, description="차원별 상세 해석 조건 목록"
    )
    override_fields: list[SearchDimension] = Field(
        default_factory=list, description="명시적 필터 파라미터로 override된 차원 이름 목록"
    )
    uninterpreted_terms: list[str] = Field(
        default_factory=list, description="구조화 조건으로 해석되지 않은 독립 단어/토큰 목록"
    )


class DimensionVerdicts(BaseModel):
    region: Literal["match", "mismatch", "unknown"] | None = Field(
        default=None, description="지역 판정 (null: 차원 미적용, match: 부합, mismatch: 불일치, unknown: 데이터 부족)"
    )
    age: Literal["match", "mismatch", "unknown"] | None = Field(
        default=None, description="연령 판정 (null: 차원 미적용, match: 부합, mismatch: 불일치, unknown: 데이터 부족)"
    )
    status: Literal["match", "mismatch", "unknown"] | None = Field(
        default=None, description="신청상태 판정 (null: 차원 미적용, match: 부합, mismatch: 불일치, unknown: 데이터 부족)"
    )
    category: Literal["match", "mismatch", "unknown"] | None = Field(
        default=None, description="카테고리 판정 (null: 차원 미적용, match: 부합, mismatch: 불일치, unknown: 데이터 부족)"
    )


class UnconfirmedCondition(BaseModel):
    field: str = Field(..., description="미확인 차원/필드명 (예: 'region', 'age')")
    reason_code: str = Field(..., description="확장 가능한 머신 판독용 미확인 사유 코드")
    message: str = Field(..., description="사용자 표시용 설명 메시지")


class PolicySearchResultItem(BaseModel):
    policy: PolicyRead
    score: float = Field(
        ...,
        description="Backend 내부 ranking 점수 (높을수록 관련도 높음, 다른 요청 간 점수 비교 불가, Release 1 UI 미표시)",
    )
    verdicts: DimensionVerdicts = Field(
        ..., description="차원별 match/mismatch/unknown/null 판정"
    )
    unknown_count: int = Field(
        ..., description="verdicts 중 null을 제외하고 unknown인 차원의 개수"
    )
    reason_codes: list[str] = Field(
        default_factory=list, description="머신 판독용 판정 이유 코드 목록 (예: ['REGION_MATCH', 'AGE_UNKNOWN'])"
    )
    message: str = Field(..., description="사용자 표시용 조건 판정 요약 메시지")
    unconfirmed_conditions: list[UnconfirmedCondition] = Field(
        default_factory=list, description="원문 데이터 미비로 인한 unknown 설명 목록"
    )


class PolicySearchResponse(BaseModel):
    total: int = Field(..., description="pagination 적용 전 필터링 조건을 만족하는 총 결과 수")
    page: int = Field(..., description="현재 페이지 번호")
    limit: int = Field(..., description="페이지당 결과 수")
    interpreted_conditions: InterpretedConditions = Field(..., description="자연어 및 명시 필터 해석 메타데이터")
    items: list[PolicySearchResultItem] = Field(default_factory=list, description="검색 결과 정책 목록")

File: frontend/src/types/policySearch.ts
/**
 * Gate G1 approved Policy Search API contract types (Frontend 04).
 *
 * Aligns with Backend 06 W3-B0 `GET /api/v1/policies/search`.
 *
 * Fixed consumption constraints:
 * - Frontend sends raw Korean `q` (trimmed, required); no Frontend NL parser.
 * - Flat query params only on the wire; no nested `structured` object.
 * - Explicit flat filters override the same dimension interpreted from `q`.
 * - Region·age·status·category verdicts: match | mismatch | unknown | null (Backend sole authority).
 * - Invalid policies are never public.
 * - Existing GET /api/v1/policies list/detail unchanged until search API ships.
 */

import type {
  ApplicationStatus,
  PolicyCategory,
  PolicyDto,
  PolicySort,
} from './policy.js';

/** Backend evaluation verdict for a searchable dimension. Aligns with Backend `MatchVerdict`. */
export type MatchVerdict = 'match' | 'mismatch' | 'unknown';

/** Dimension key for NL interpretation and per-policy verdicts. */
export type InterpretedConditionDimension =
  | 'keyword'
  | 'region'
  | 'age'
  | 'category'
  | 'status';

/** Whether a condition came from natural-language `q` or an explicit flat filter. */
export type InterpretedConditionSource = 'q' | 'explicit';

/** Backend resolution for a parsed condition dimension. */
export type InterpretedConditionResolution = 'resolved' | 'unmapped' | 'ambiguous';

/**
 * Resolved value for one interpreted condition.
 * Shape depends on `dimension` (keyword/region string, age number, category/status enum).
 */
export type InterpretedConditionValue =
  | string
  | number
  | PolicyCategory
  | ApplicationStatus;

/** One row in response `interpreted_conditions.conditions`. */
export interface InterpretedCondition {
  dimension: InterpretedConditionDimension;
  value: InterpretedConditionValue;
  source: InterpretedConditionSource;
  resolution: InterpretedConditionResolution;
  /** Candidate strings when resolution is `ambiguous` (e.g. region alias matches). */
  candidates: string[];
}

/** Top-level NL interpretation block on {@link PolicySearchResponse}. */
export interface PolicySearchInterpretedConditions {
  q_raw: string;
  q_clean: string;
  conditions: InterpretedCondition[];
  /** Dimensions where explicit flat params overrode `q` interpretation. */
  override_fields: InterpretedConditionDimension[];
  /** Tokens from `q` that could not be mapped to a searchable dimension. */
  uninterpreted_terms: string[];
}

/**
 * Per-policy dimension verdicts returned by Backend search evaluation.
 * `null` = condition not applied to this search; `unknown` = applied but no policy evidence.
 */
export interface DimensionVerdicts {
  region: MatchVerdict | null;
  age: MatchVerdict | null;
  status: MatchVerdict | null;
  category: MatchVerdict | null;
}

/**
 * Machine-readable reason code from Backend evaluation.
 * Extensible string — unknown codes must not break UI (see Forest plan § Reason fallback).
 */
export type ReasonCode = string;

/** Per-row dimension still unconfirmed for this policy (not guessed as match). */
export interface UnconfirmedCondition {
  field: string;
  reason_code: ReasonCode;
  message: string;
}

/**
 * GET /api/v1/policies/search flat query parameters.
 * Mirrors Backend Pydantic request model field names, nullability, and defaults.
 */
export interface PolicySearchQueryParams {
  /** Natural-language query. Required after trim; empty → 422. Max 200 chars. */
  q: string;
  /** Optional keyword text filter. Max 100 chars. */
  keyword?: string | null;
  /** Administrative region alias or canonical name string. Max 100 chars. */
  region?: string | null;
  age?: number | null;
  category?: PolicyCategory | null;
  status?: ApplicationStatus | null;
  /** Default `true` when omitted (Backend and Frontend URL state). */
  include_partial?: boolean;
  page?: number;
  limit?: number;
  sort?: PolicySort;
  /** Browser-only profile preferences sent in a POST body, never as URL filters. */
  preferences?: PolicySearchPreferences | null;
}

export interface PolicySearchPreferences {
  region?: string | null;
  age?: number | null;
  categories: PolicyCategory[];
}

/** Backend defaults — keep in sync with Backend 06 W3-B0. */
export const POLICY_SEARCH_DEFAULTS = {
  include_partial: true,
  page: 1,
  limit: 20,
  sort: 'default',
} as const;

/** Recommended query string length limits (422 when exceeded). */
export const POLICY_SEARCH_QUERY_LIMITS = {
  q: 200,
  keyword: 100,
  region: 100,
} as const;

/**
 * One ranked search hit: nested PolicyRead (`PolicyDto`) plus search metadata.
 * Matches Backend nested DTO — not a flat merge of PolicyRead fields.
 *
 * `score` is for Backend ordering only — Release 1 UI does not display it as a number
 * and must not compare scores across different search requests.
 */
export interface PolicySearchHit {
  policy: PolicyDto;
  score: number;
  verdicts: DimensionVerdicts;
  /** Number of applied verdict dimensions whose result is `unknown`. */
  unknown_count: number;
  reason_codes: ReasonCode[];
  message: string;
  unconfirmed_conditions: UnconfirmedCondition[];
}

/** Search response envelope. Pagination matches PolicyListResponse shape. */
export interface PolicySearchResponse {
  total: number;
  page: number;
  limit: number;
  interpreted_conditions: PolicySearchInterpretedConditions;
  items: PolicySearchHit[];
}

/** Approved search endpoint (G1 integration). */
export const POLICY_SEARCH_ENDPOINT = {
  method: 'GET' as const,
  preferenceMethod: 'POST' as const,
  path: '/api/v1/policies/search' as const,
};
